func_darwin.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2022 The Ebitengine Authors
+
+package purego
+
+import (
+	"reflect"
+	"runtime"
+	"unsafe"
+
+	"github.com/ebitengine/purego/internal/strings"
+)
+
+// FuncHandle takes a handle to a shared object returned from Dlopen, the name of a C function in that
+// shared object and a pointer to a Go function representing the calling convention of the C function.
+// fptr will be set to a function that when called will call the C function given by name with the
+// parameters passed in the correct registers and stack.
+//
+// A panic is produced if the name symbol cannot be found in handle or if the type is not a function
+// pointer or if the function returns more than 1 value.
+//
+// These conversions describe how a Go type in the fptr will be used to call
+// the C function. It is important to note that there is no way to verify that fptr
+// matches the C function. This also holds true for struct types where the padding
+// needs to be ensured to match that of C; FuncHandle does not verify this.
+//
+// Conversion Type (Go => C)
+//
+//	string => char*
+//	bool => _Bool
+//	uintptr => uintptr_t
+//	uint => System Dependent
+//	uint8 => uint8_t
+//	uint16 => uint16_t
+//	uint32 => uint32_t
+//	uint64 => uint64_t
+//	int => System Dependent
+//	int8 => int8_t
+//	int16 => int16_t
+//	int32 => int32_t
+//	int64 => int64_t
+//	float32 => float
+//	float64 => double
+//	struct => struct
+//	func => C function
+//	[]T, unsafe.Pointer, *T => void*
+//
+// There is a special case when the last argument of fptr is a variadic interface
+// it will be expanded into a call to the C function as if it had those arguments.
+func FuncHandle(handle uintptr, name string, fptr interface{}) {
```

Comment:
I feel like this function name should start with a verb like `Register`, but what do you think?

Reply:
`RegisterFuncHandle` seems good. Kind of long though

Reply:
Hmmm Should it actually be `RegisterHandleFunc` since in English the adjective usually goes on the right of the noun it describes. In this case Handle describes the type of function it is. Does that make sense?

Reply:
Or just `RegisterFunc`?

Reply:
That could work. However, `RegisterFuncHandle` specifically takes a handle to a shared object while the unexported function `registerFunc`  takes the C function. There may be a time that exporting that function might make sense. It could be useful while porting C code. Something like this:

```
// int foo(char *c) { /* ... */ }
import "C"
var fooFn func(string) int
purego.RegisterFunc(C.foo, &fooFn)
```
This way you don't have to port the entire C function before having something running.

If you don't think that's a convincing argument than let's just go with `RegisterFunc`. Thoughts?

Reply:
I see. What about

* `FuncHandle` → `RegisterFuncFromLib`
* `_Func` → `RegisterFunc`

? The former might be too verbose...

```go
func init() {
    purego.RegisterFuncFromLib(objc, "objc_msgSend", &objc_msgSend)
}
```

Even without the former, we can write like this:

```go
func init() {
    purego.RegisterFunc(Dlsym(objc, "objc_msgSend"), &objc_msgSend)
}
```

I slightly prefer only having RegisterFunc.

Reply:
That is correct. The function is the destination and handle and name are source arguments

Reply:
OK so I'd be happy if you could swap the arguments.

Reply:
RegisterHandleFunc is just a wrapper around registerFunc which can be added later if it is useful. I'm kind of thinking tho that most people will do what you wrote and not check that the Dlopen succeeded or not. RegisterHandleFunc will at least force it

Reply:
I see. Needing to check errors makes sense. Another suggestion is `RegisterLibFunc`:

```go
func init() {
    purego.RegisterLibFunc(&objc_msgSend, objc, "objc_msgSend")
}
```

Reply:
I like it! If that way if we decide to export `RegisterFunc` we can.

Reply:
I'm fine to expose RegisterFunc later, but we don't have to do it now :-)

Reply:
Confirmed RegisterFunc was exported. That's also fine :-)